runtime/src/bank/builtins/core_bpf_migration/mod.rs

```diff
@@ -54,84 +55,71 @@ pub(crate) struct CoreBpfMigrationConfig {
-        programdata_address: target.program_data_address,
-    };
-    let data = bincode::serialize(&state)?;
-    // The source program account has the same state, so it should already have
-    // a sufficient lamports balance to cover rent for this state.
-    // Out of an abundance of caution, first ensure the source program
-    // account's data is the same length as the serialized state.
-    if source.program_account.data().len() != data.len() {
-        return Err(CoreBpfMigrationError::InvalidProgramAccount(
-            source.program_address,
-        ));
-    }
-    // Then copy the source account's contents and overwrite the data with the
-    // newly created target program account data.
-    let mut account = source.program_account.clone();
-    account.set_data_from_slice(&data);
-    Ok(account)
+fn checked_sub<T: CheckedSub>(a: T, b: T) -> Result<T, CoreBpfMigrationError> {
+    a.checked_sub(&b)
+        .ok_or(CoreBpfMigrationError::ArithmeticOverflow)
 }
 
-/// Create an `AccountSharedData` with data initialized to
-/// `UpgradeableLoaderState::ProgramData` populated with the current slot, as
-/// well as the source program data account's upgrade authority and ELF.
-///
-/// Note that the account's data is initialized manually, but the rest of the
-/// account's fields are inherited from the source program account, including
-/// the lamports.
-fn new_target_program_data_account(
-    source: &SourceUpgradeableBpf,
-    slot: Slot,
-) -> Result<AccountSharedData, CoreBpfMigrationError> {
-    let programdata_data_offset = UpgradeableLoaderState::size_of_programdata_metadata();
-    // Deserialize the program data metadata to get the upgrade authority.
-    if let UpgradeableLoaderState::ProgramData {
-        upgrade_authority_address,
-        ..
-    } = bincode::deserialize(&source.program_data_account.data()[..programdata_data_offset])?
-    {
-        let mut account = source.program_data_account.clone();
-        // This account's data was just partially deserialized into
-        // `UpgradeableLoaderState`, so it's guaranteed to have at least enough
-        // space for the same type to be serialized in.
-        // The ELF should remain untouched, since it follows the
-        // `UpgradeableLoaderState`.
-        //
-        // Serialize the new `UpgradeableLoaderState` with the bank's current
-        // slot and the deserialized upgrade authority.
-        bincode::serialize_into(
-            account.data_as_mut_slice(),
-            &UpgradeableLoaderState::ProgramData {
-                slot,
-                upgrade_authority_address,
-            },
-        )?;
-        return Ok(account);
-    }
-    Err(CoreBpfMigrationError::InvalidProgramDataAccount(
-        source.program_data_address,
-    ))
-}
+const BUFFER_METADATA_SIZE: usize = UpgradeableLoaderState::size_of_buffer_metadata();
+const PROGRAMDATA_METADATA_SIZE: usize = UpgradeableLoaderState::size_of_programdata_metadata();
```

Comment:
I know you two had a lot of back and forth about constants, but I don't see that these contribute a lot, except maybe reduce compile time a tiny amount. Otherwise, they just make the reader have to learn new terminology beyond the UpgradeableLoaderState api that already exists.

Reply:
I think @Lichtso's point was moreso around using the `UpgradeableLoaderState` API, not necessarily the constants. If that's accurate I can pop them.

Reply:
Popped for now.